Tuya Cover improvements

pauln opened the pull request:
# What does this implement/fix? 

This PR reworks #1113 based on #2279, which has been merged.

Adds support for:
- Restore mode
- Control datapoint (for open/close/stop commands)
- Position report datapoint (if the device reports position and accepts position commands on separate datapoints)
- Setting device direction-of-travel if the `invert_position` option is enabled and a `direction_datapoint` is configured

Note that the `invert_position` option is no longer used to determine whether to invert the values, as it appears that the Tuya cover position is the inverse of the HA / ESPHome position (percentage open vs percentage closed).  This was discussed briefly [in the comments on my earlier PR](https://github.com/esphome/esphome/pull/1113#issuecomment-951714160) in an effort to avoid any breaking changes.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1578

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

uart:
  rx_pin: GPIO13
  tx_pin: GPIO15
  baud_rate: 9600
  
tuya:

cover:
  - platform: tuya
    name: "Lounge curtains"
    device_class: curtain
    control_datapoint: 1
    direction_datapoint: 5
    position_datapoint: 2
    position_report_datapoint: 3
    invert_position: True
    restore_mode: RESTORE_AND_CALL
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

pauln committed:
Tuya: Add force_set_*_datapoint functions

Adds a set of functions to "force set" datapoints of the various types (sending the new value even if it doesn't seem to differ from the current value).  These are useful for datapoints which trigger an action but which don't necessarily set/retain their value properly.

pauln committed:
Tuya Cover: Add support for more features

Adds support for:
- Restore mode
- Control datapoint (for open/close/stop commands)
- Position report datapoint (if the device reports position and accepts position commands on separate datapoints)
- Setting device direction-of-travel if the `invert_position` option is enabled and a `direction_datapoint` is configured

Note that the `invert_position` option is no longer used to determine whether to invert the values, as it appears that the Tuya cover position is the inverse of the HA / ESPHome position (percentage open vs percentage closed).

pauln committed:
Appease clang-format

pauln committed:
Align function name in tuya_cover.h

pauln committed:
Finish renaming set_direction_() in tuya_cover.cpp

It sometimes helps to wait for the rest of the CI steps to finish, in case they provide more detail...

pauln committed:
Tuya Cover: Position datapoint is required, so don't check if it's set

We don't need to check if the position datapoint has been configured, as it's required.  This should stop the clang-tidy ESP32 4/4 CI step from failing.

pauln committed:
Tuya Cover: Simplify listener registration

Since the main position datapoint ID is required, we can use it as the default rather than checking whether either datapoint ID is configured.

pauln commented:
What should happen with `RESTORE_AND_CALL` is that the state is restored (in HA) to the previous position and the Tuya MCU is instructed to move to that position.  I'm not sure that setting the position in HA after the device is unplugged would be expected to work, though, as I think the restore happens from the device's flash rather than from HA - so a better test might be to unplug the device, manually move the cover, then plug it back in and see if it returns to where it was before you unplugged it.

I'd previously only tested `RESTORE_AND_CALL` by watching for it sending the commands on restart, but I've just given this "move the cover while the device is unplugged" approach and found that it doesn't really work (for my motors) - the position sensing must be relative rather than absolute, so it's not able to tell that it's moved while unplugged and so thinks it's already in the requested place.  It therefore put itself into an error / uncalibrated state and required recalibrating by closing and opening to let it find the end positions before it could correctly report or move to a position.  I'm not sure whether this is reflective of all Tuya cover devices, though - so if you're willing to try this out (with the understanding that you'll need to recalibrate it if its position sensing isn't absolute), that'll help us figure out whether `RESTORE_AND_CALL` is likely to make sense for any Tuya cover device (if not, I might remove it from the supported restore modes).  I'll add a note to the docs about it either way, to avoid giving a false impression of what the device is actually capable of.

oxan commented on esphome/components/tuya/cover/tuya_cover.h:
```diff
@@ -7,26 +7,47 @@
 namespace esphome {
 namespace tuya {
 
+const extern uint8_t COMMAND_OPEN;
+const extern uint8_t COMMAND_CLOSE;
+const extern uint8_t COMMAND_STOP;
```
These declarations seems unnecessary.

pauln replied:
Do you mean that these don't need to be declared in the header file (but can/should still be in the `.cpp` file), or that you don't think these are worth declaring as consts at all?  If it's that they don't need to be declared in the header file, thanks - I don't do a lot of programming in C(++), so I try to follow patterns I see in the existing code around aspects I'm not sure on.  If you think the consts themselves are unnecessary, I think they improve the readability of the code - but since they're not used a whole lot, I could remove them (and add comments to clarify the "magic numbers") if you'd prefer.

oxan replied:
> If it's that they don't need to be declared in the header file, thanks - I don't do a lot of programming in C(++), so I try to follow patterns I see in the existing code around aspects I'm not sure on.

Yes, that. `extern` declarations are only needed if a constant defined in one (.cpp) file is used in another, which doesn't seem to be the case here. I agree having them as constants instead of using magic numbers is preferable.

oxan commented on esphome/components/tuya/cover/tuya_cover.cpp:
```diff
@@ -4,48 +4,129 @@
 }
 
+void TuyaCover::loop() {
+  // Wait until the Tuya component reports state INIT_DONE before attempting
+  // to set the direction of travel or restore the state (if applicable), as
+  // these won't work if the Tuya MCU isn't ready yet.
```
I think having an `on_initialized` callback or something on the Tuya component would be a nicer solution.

pauln replied:
An `on_initialized` callback would indeed be nice, but I'm not entirely sure how to go about implementing such a thing.  I'm happy to shift this out of the `loop()` if somebody who's more familiar with that sort of thing is willing to implement the callback process.

oxan replied:
In the `Tuya` class, add a `CallbackManager<void()> initialized_callback_{}` member, and a function:
```cpp
void add_on_initialized_callback(std::function<void()> callback) {
  this->initialized_callback_.add(std::move(callback));
}
```
Then, when it's initialized, it can execute the callback with `this->time_sync_callback_.call()`.

In the setup of the cover you can then do something like this:
```
this->parent_->add_on_initialized_callback([this]() {
  // handle restore here
});
```

pauln replied:
That's much simpler than I'd imagined it to be - thanks!  I've switched over to this, which is indeed a much cleaner solution.

pauln committed:
Tuya Cover: Remove unnecessary const extern declarations

pauln committed:
Tuya Cover: Implement a callback instead of doing init in the loop

Implements an on_initialized callback in the main Tuya component, and moves Tuya Cover init (restore mode and direction config) into the callback rather than tracking the ready state in the loop.  Thanks to @oxan for revealing how easy to implement callbacks are!

oxan commented on esphome/components/tuya/tuya.h:
```diff
@@ -81,12 +82,22 @@ class Tuya : public Component, public uart::UARTDevice {
+  void force_set_string_datapoint_value(uint8_t datapoint_id, const std::string &value);
+  void force_set_enum_datapoint_value(uint8_t datapoint_id, uint8_t value);
+  void force_set_bitmask_datapoint_value(uint8_t datapoint_id, uint32_t value, uint8_t length);
+  TuyaInitState get_init_state();
```
This method seems superfluous now.

oxan reviewed (approved):
LGTM, though I'd prefer a review from someone a bit more familar with Tuya as well.

jesserockz commented on esphome/components/tuya/cover/tuya_cover.cpp:
```diff
@@ -4,48 +4,122 @@
+
+  this->parent_->add_on_initialized_callback([this]() {
+    // Set the direction (if configured/supported).
+    this->set_direction_(this->invert_position_);
+
```
How can someone invert the position without a direction datapoint now?

pauln replied:
Re-reading Marek's comments, it seems as though the direction datapoint is supported on their device even though the datapoint may not have been advertised (similarly to the control datapoint), and I wouldn't be surprised if it's actually an integral part of the Tuya cover device spec (if such a thing exists).  As such, I'm not sure it's worth the extra complexity to support "soft" inversion at this point - but we could certainly add it back in the future if anyone comes across a device which doesn't support the datapoint but which can require inversion.

I can add a note to the docs to explain this change, if you're happy with this approach (noting that if we _do_ add back the "soft" inversion, it'd actually cause existing configs to flip since the `inverted: true` config was required to compensate for the Tuya cover positions being inverted relative to HA/ESPHome).  It might also be worth noting it in the release notes.
